ui/app/src/Views/AlertList/FilterBar.elm

```diff
@@ -1,20 +1,137 @@
 module Views.AlertList.FilterBar exposing (view)
 
-import Html exposing (Html, div, span, input, text, button, i)
-import Html.Attributes exposing (value, class)
-import Html.Events exposing (onClick, onInput)
+import Html exposing (Html, Attribute, div, span, input, text, button, i, small)
+import Html.Attributes exposing (value, class, style, disabled, id)
+import Html.Events exposing (onClick, onInput, on, keyCode)
+import Utils.Filter exposing (Matcher)
+import Views.AlertList.Types exposing (AlertListMsg(..))
+import Types exposing (Msg(..))
+import Json.Decode as Json
 
 
-view : String -> (String -> msg) -> msg -> Html msg
-view filterText inputChangedMsg buttonClickedMsg =
-    div [ class "input-group" ]
-        [ span [ class "input-group-addon" ]
-            [ i [ class "fa fa-filter" ] []
+viewMatcher : Matcher -> Html Msg
+viewMatcher matcher =
+    div [ class "col col-auto", style [ ( "padding", "5px" ) ] ]
+        [ div [ class "btn-group" ]
+            [ button
+                [ class "btn btn-outline-info"
+                , onClick (DeleteFilterMatcher True matcher |> MsgForAlertList)
+                ]
+                [ text <| Utils.Filter.stringifyMatcher matcher
+                ]
+            , button
+                [ class "btn btn-outline-danger"
+                , onClick (DeleteFilterMatcher False matcher |> MsgForAlertList)
+                ]
+                [ text "×" ]
             ]
-        , input
-            [ class "form-control", value filterText, onInput inputChangedMsg ]
-            []
-        , span
-            [ class "input-group-btn" ]
-            [ button [ class "btn btn-primary", onClick buttonClickedMsg ] [ text "Filter" ] ]
         ]
+
+
+lastElem : List a -> Maybe a
+lastElem =
+    List.foldl (Just >> always) Nothing
+
+
+viewMatchers : List Matcher -> List (Html Msg)
+viewMatchers matchers =
+    matchers
+        |> List.map viewMatcher
+
+
+onKey : String -> Int -> Msg -> Attribute Msg
+onKey event key msg =
+    on event
+        (Json.map
+            (\k ->
+                if k == key then
+                    msg
+                else
+                    Noop
+            )
+            keyCode
+        )
+
+
+view : List Matcher -> String -> Bool -> Html Msg
+view matchers matcherText backspacePressed =
+    let
+        className =
+            if matcherText == "" then
+                ""
+            else
+                case maybeMatcher of
+                    Just _ ->
+                        "has-success"
+
+                    Nothing ->
+                        "has-danger"
+
+        maybeMatcher =
+            Utils.Filter.parseMatcher matcherText
+
+        onKeydown =
+            onKey "keydown" 8 <|
+                case ( matcherText, backspacePressed ) of
```

Comment:
We had a look at it together today. Holding down the backspace removes (one by one) each last character in the input, and the whole time sends multiple keyDown events. This is a guard so that if a user holds down backspace to remove the text in the input, they won't accidentally hold backspace too long and then delete the preceding matcher as well. So, once a user holds backspace to clear an input, they have to then lift up the key and press it again to proceed to deleting the next matcher. I think this makes sense to prevent users from deleting all their matchers at once accidentally.